pyatv/core/scan.py

```diff
@@ -480,3 +456,203 @@ async def process(self, timeout: int) -> None:
+        for info in infos:
+            if info.type == DEVICE_INFO_TYPE:
+                model = info.properties.get(b"model")
+                if model:
+                    name = _extract_service_name(info)
+                    with contextlib.suppress(UnicodeDecodeError):
+                        name_to_model[name] = model.decode("utf-8")
+            else:
+                addresses = cast(
+                    List[IPv4Address], info.ip_addresses_by_version(IPVersion.V4Only)
+                )
+                for ip_address in addresses:
+                    if hosts and ip_address not in hosts:
+                        continue
+                    services_by_address.setdefault(ip_address, []).append(info)
+        return services_by_address, name_to_model
+
+
+class ZeroconfMulticastScanner(ZeroconfScanner):
+    """Service discovery using zeroconf.
+
+    A ServiceBrowser must be running for all the types we are browsing.
+    """
+
+    async def _lookup_services_and_models(
+        self, zc_timeout: float
+    ) -> Tuple[Dict[IPv4Address, List[AsyncServiceInfo]], Dict[str, str]]:
+        """Lookup services and aggregate them by address."""
+        infos = self._build_service_info_queries()
+        requests: List[Awaitable[bool]] = [
+            info.async_request(self.zeroconf, zc_timeout)
+            for info in infos
+            if not info.load_from_cache(self.zeroconf)
+        ]
+        if requests:
+            await asyncio.gather(*requests)
+        return self._process_service_info_responses(infos)
+
+
+class ZeroconfUnicastScanner(ZeroconfScanner):
+    """Service discovery using zeroconf.
+
+    A ServiceBrowser should be running for all the types we are browsing,
+    but we will fallback to sending unicast PTR queries for missing types
+    in the event the network is dropping multicast responses for some
+    reason.
+    """
+
+    def __init__(
+        self,
+        aiozc: AsyncZeroconf,
+        hosts: List[IPv4Address],
+    ) -> None:
+        """Initialize a new scanner."""
+        super().__init__(aiozc, hosts)
+        hosts = self.hosts
+        # Once info_by_address_type is filled in, we are done.
+        self.infos_by_address_type: Dict[
+            IPv4Address, dict[str, Optional[AsyncServiceInfo]]
+        ] = {host: {f"{type_}.": None for type_ in self._services} for host in hosts}
+
+    def _send_ptr_queries(
+        self, address: IPv4Address, needed_types: List[str]
+    ) -> None:
+        """Send PTR queries."""
+        out = DNSOutgoing(_FLAGS_QR_QUERY)
+        # Note that the multicast flag here is True even though we are sending
+        # unicast queries. We still are speaking mdns and will be happy to get
+        # a multicast response back.
+        for type_ in needed_types:
+            question = DNSQuestion(type_, _TYPE_PTR, _CLASS_IN)
+            question.unicast = True
+            out.add_question(question)
+        self.zeroconf.async_send(out, str(address))
+
```

Comment:
This is a bit of a big hammer but its unicast to a single host so only HA has to pay to process it and we aren't flooding the network with requests so it seems like an ok trade off